flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java

```diff
@@ -0,0 +1,255 @@
+                            < materializationID) {
+                        LOG.info(
+                                "Update checkpoint {}, old materializationID {}, new materializationID {}.",
+                                checkpointId,
+                                mapToMaterializationId.get(checkpointId),
+                                materializationID);
+                    }
+                    mapToMaterializationId.put(checkpointId, materializationID);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void storeLocalState(long checkpointId, @Nullable TaskStateSnapshot localState) {
+        if (checkpointId < lastCheckpointId) {
+            LOG.info(
+                    "Current checkpoint {} is out of order, smaller than last CheckpointId {}.",
+                    lastCheckpointId,
+                    checkpointId);
+            return;
+        } else {
+            lastCheckpointId = checkpointId;
+        }
+        synchronized (lock) {
+            updateReference(checkpointId, localState);
+        }
+        super.storeLocalState(checkpointId, localState);
+    }
+
+    @Override
+    protected File getCheckpointDirectory(long checkpointId) {
+        final File checkpointDirectory =
+                localRecoveryConfig
+                        .getLocalStateDirectoryProvider()
+                        .orElseThrow(
+                                () -> new IllegalStateException("Local recovery must be enabled."))
+                        .subtaskBaseDirectory(checkpointId);
+        File directoryForChangelog =
+                new File(checkpointDirectory, CHANGE_LOG_CHECKPOINT_PREFIX + checkpointId);
+
+        if (!directoryForChangelog.exists() && !directoryForChangelog.mkdirs()) {
+            throw new FlinkRuntimeException(
+                    String.format(
+                            "Could not create the checkpoint directory '%s'",
+                            directoryForChangelog));
+        }
+
+        return directoryForChangelog;
+    }
+
+    private void deleteMaterialization(LongPredicate pruningChecker) {
+        final Set<Long> materializationToRemove = new HashSet<>();
+        synchronized (lock) {
+            Iterator<Entry<Long, Long>> iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                long entryCheckpointId = entry.getKey();
+                if (pruningChecker.test(entryCheckpointId)) {
+                    materializationToRemove.add(entry.getValue());
+                    iterator.remove();
+                }
+            }
+
+            iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                materializationToRemove.remove(entry.getValue());
+            }
```

Comment:
How about simplifying this code?
```
Set<Long> checkpoints =
        mapToMaterializationId.keySet().stream()
                .filter(pruningChecker::test)
                .collect(Collectors.toSet());
materializationToRemove =
        checkpoints.stream()
                .map(mapToMaterializationId::remove)
                .collect(Collectors.toSet());
materializationToRemove.removeAll(mapToMaterializationId.values());
```

Alternatively, we could maintain a Map<materializationId, Set<checkpointId>> and discard materialization whenever the set is empty. But two maps seems more complex.